lsp-mode.el

```diff
@@ -6986,6 +6998,38 @@ nil."
     nil))
 
 
+;; Download URL handling
+(cl-defun lsp-download-install (callback error-callback &key url store-path &allow-other-keys)
+  (let ((url (lsp-resolve-value url))
+        (store-path (lsp-resolve-value store-path)))
+    (make-thread
+     (lambda ()
+       (condition-case err
+           (progn
+             (when (f-exists? store-path)
+               (f-delete store-path))
+             (lsp--info "Starting to download %s to %s..." url store-path)
+             (mkdir (f-parent store-path) t)
+             (url-copy-file url store-path)
+             (lsp--info "Finished downloading %s..." store-path)
+             (funcall callback))
+         (error (funcall error-callback err)))))))
+
+(cl-defun lsp-download-path (&key store-path set-executable? &allow-other-keys)
+  "Download URL and store it into STORE-PATH.
+
+SET-EXECUTABLE? when non-nil change the executable flags of
+STORE-PATH to make it executable."
+  (let ((store-path (lsp-resolve-value store-path)))
+    (cond
+     ((executable-find store-path) store-path)
```

Comment:
In Windows, given a `store-path` without extension for `executable-find`, it can be extended to full path with extension (`.exe`, `.cmd`, ...).
I think you should put this one out of `cond` as an optional resolvers  like `(let ((store-path (or (executable-find store-path) store-path))))` and do the `cond` with the rest conditional switch.

Reply:
I am not sure I understand the comment 100%. Are you suggesting to perform executable-find two times in this function?

Reply:
No, you will only need to perform executable-find once.

``` elisp
(let* ((store-path (lsp-resolve-value store-path))
       (store-path (or (executable-find store-path) store-path)))
  (cond
   ((and set-executable? (f-exists? store-path))
    (set-file-modes store-path #o0700)
    store-path)
   ((f-exists? store-path) store-path)))
```

Reply:
Structuring the code this way will result it calling set-file-modes everytime.

Reply:
Does setting file mode every time cause problem?
If that's not what you want, I think we can always temporarily cache the value from `executable-path` to prevent it form calling twice.

In Windows, if you have file `test`, `test.cmd` in the same folder and, the `executable-find` will return `test.cmd` and this function will return `test`. Which is not executable by Windows.

Reply:
It is syscall so it is slow and it will be better if we can avoid it. I am looking at your suggestion and in the code in the PR and I fail to see a case in which both will act differently. Can you give an example?

Reply:
For example, if you have `a.cmd` under `~/tmp/`.
In Windows, with your original function:
``` elisp
(lsp-download-path :store-path "~/tmp/a")
"~/tmp/a"
```
and with the modified function
``` elisp
(lsp-download-path :store-path "~/tmp/a")
"c:/Users/xyz/tmp/a.cmd"
```

Since `~/tmp/a` is not existed, shouldn't the function return `nil` or correct path with suffix to executable file?

My previous suggestion is under assumption that you want to return the correct path with suffix.
But if that's not correct and you actually want to return `nil` here, maybe you should use the combination of `file-executable-p` and `file-directory-p` instead of `executable-find`, which will work on exact `store-path`.

Reply:
> My previous suggestion is under assumption that you want to return the correct path with suffix.

There is no such need - this will be handled later by the code that will run the server.

Reply:
> There is no such need - this will be handled later by the code that will run the server.

Do you mean the code in `:new-connection` will do that?
Another extreme example, you're suppose to download `a.exe` to `bin/` folder, and there's already a `a.exe.cmd` file under that folder.
Now if your download failed somehow, the `lsp-download-path` will still return `bin/a.exe` (since `a.exe.cmd` is executable) and the server will fail to start. Isn't that a bug?

Reply:
> Do you mean the code in :new-connection will do that?

The code in this PR does the same thing that we do in  `lsp-server-present?`.

> Now if your download failed somehow, the `lsp-download-path` will still return `bin/a.exe` (since `a.exe.cmd` is executable) and the server will fail to start. Isn't that a bug?

That same thing will happen if you set the server to `C:\server.exe` and it does not exist but there is `C:\server.exe.cmd`. I don't think that we should spend effort on handling cases like that.

Reply:
> The code in this PR does the same thing that we do in `lsp-server-present?`.

Is that the same? `lsp-server-present?` is returning `(executable-find path)` result instead `path` though.

> That same thing will happen if you set the server to `C:\server.exe` and it does not exist but there is `C:\server.exe.cmd`. I don't think that we should spend effort on handling cases like that.

But we normally just set the server to `server` so it works in all platform, don't we?

I think the check is kind of a bug and user may be affected by it. It can be easily resolved with one line change, why shouldn't we fix it?

Reply:
> Is that the same? `lsp-server-present?` is returning `(executable-find path)` result instead `path` though.

You are missing the point. `lsp-server-present?` returns `nil`/`non-nil` as an indication that the server is present. We do the same in this function. 

> I think the check is kind of a bug and user may be affected by it. It can be easily resolved with one line change, why shouldn't we fix it?

The chances that anyone will create `rust-analyzer.exe.cmd` and we will mistakenly start the server. I don't understand what is the bug that your suggestion will fix.

Reply:
> You are missing the point. `lsp-server-present?` returns `nil`/`non-nil` as an indication that the server is present. We do the same in this function.

Except that in this function we even return `non-nil` for non existed server?
Since currently `lsp-server-present?` takes output from command lambda, which is using `lsp-download-path`, if `lsp-download-path` is returning `nil` for non existed server, that would make it fallback to final condition clause and allow `lsp-server-present?` takes correct parameters.



> The chances that anyone will create `rust-analyzer.exe.cmd` and we will mistakenly start the server. I don't understand what is the bug that your suggestion will fix.

With my suggestion in https://github.com/emacs-lsp/lsp-mode/pull/2197#discussion_r493687753, we will not mistakenly start the server even when `rust-analyzer.exe` is not existed.


The discussion has gotten long. If you think my point is invalid then it's fine, I'm ok with that.

Reply:
Sorry I got lost. If the whole discussion is to guard us against user creating `~/.emacs.d/.cache/lsp-rust/rust-analyzer.exe.cmd` which then we will mistakenly assume as the rust-analyzer then the discussion shouldn't have started at all because this is not something that ever going to happen and even if it happens it will be entirely users fault.

Reply:
> it will be entirely users fault

Even if it's user fault, it will be a special case since user's `lsp-mode` is not working.
Further more, the error message in `*lsp-log*` doesn't mention anything like that, so it will be hard for user to know why their configuration is not working.

My main concern in this function is we check for one thing (`(executable-find path)`) but returns another thing `path`, which seems weird.

I've approved the PR anyway.